tests/buffering.rs

```diff
@@ -71,17 +70,16 @@ fn test_buffering() {
         .send_all(input_events_stream);
     let _ = rt.block_on(send).unwrap();
 
-    // There was a race caused by a channel in the mock source, and this
-    // check is here to ensure it's really gone.
-    assert_eq!(source_event_counter.load(Ordering::Acquire), num_events);
-
-    // Give the topology some time to process the received data and simulate
-    // a crash.
+    // We need to wait for at least the source to process events.
+    wait_for(|| source_event_counter.load(Ordering::Acquire) == num_events);
```

Comment:
So, this means that the changes from  https://github.com/timberio/vector/commit/e0761a9681da6b7e8cb69c750c41a7be89d24fed  don't fix the issue :( 
Can we restore the comment that was there before the said comment? I think it's more on-point, and gives hints on where the root cause of the issue is. Unless it's not relevant anymore.

Reply:
It fixed races relevant to the PR, but there is another race in MockSourceConfig that we can't avoid without changing the channel impl to some that doesn't use buffering, but even then just source processing events isn't enough, they still need to reach the buffers of source and yet still they need to reach file to which we are buffering those events. Only some intrusive synchronizing could fix this.

Because of the first race, the assert isn't appropriate.

Reply:
Yeah! The code before https://github.com/timberio/vector/commit/e0761a9681da6b7e8cb69c750c41a7be89d24fed was:

```rust
    // A race caused by `rt.block_on(send).unwrap()` is handled here. For some
    // reason, at times less events than were sent actually arrive to the
    // `source`.
    // We mitigate that by waiting on the event counter provided by our source
    // mock.
    test_util::wait_for_atomic_usize(source_event_counter, |x| x == num_events);
```

It's pretty much what you said, so I suggest just restoring it as it was.

Reply:
So, to improve the overall situation, we have to reimplement the `MockSourceConfig`, right?

Reply:
Ok, I'll document the race for because of which we need to wait. And use that `test_util::wait_for_atomic_usize`.


> So, to improve the overall situation, we have to reimplement the MockSourceConfig, right?

I don't think its possible to get better than waiting for `source_event_counter`.